oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/xml/ImporterImpl.java

```diff
@@ -336,138 +339,150 @@ public void startNode(@NotNull NodeInfo nodeInfo, @NotNull List<PropInfo> propIn
-                    // resolve uuid conflict
-                    tree = resolveUUIDConflict(parent, conflicting, id, nodeInfo);
-                    if (tree == null) {
-                        // no new node has been created, so skip this node
-                        parents.push(null); // push null onto stack for skipped node
-                        log.debug("Skipping existing node {}", nodeInfo.getName());
-                        return;
+            if (tree == null) {
+                // create node
+                if (id == null) {
+                    // no potential uuid conflict, always add new node
+                    tree = createTree(parent, nodeInfo, null);
+                } else if (uuidBehavior == ImportUUIDBehavior.IMPORT_UUID_CREATE_NEW) {
+                    // always create a new UUID even if no
+                    // conflicting node exists. see OAK-1244
+                    tree = createTree(parent, nodeInfo, UUID.randomUUID().toString());
+                    // remember uuid mapping
+                    if (isNodeType(tree, JcrConstants.MIX_REFERENCEABLE)) {
+                        refTracker.put(nodeInfo.getUUID(), TreeUtil.getString(tree, JcrConstants.JCR_UUID));
                     }
                 } else {
-                    // create new with given uuid
-                    tree = createTree(parent, nodeInfo, id);
+
+                    Tree conflicting = idLookup.getConflictingTree(id);
+                    if (conflicting != null && conflicting.exists()) {
+                        // resolve uuid conflict
+                        tree = resolveUUIDConflict(parent, conflicting, id, nodeInfo);
+                        if (tree == null) {
+                            // no new node has been created, so skip this node
+                            parents.push(null); // push null onto stack for skipped node
+                            log.debug("Skipping existing node {}", nodeInfo.getName());
+                            return;
+                        }
+                    } else {
+                        // create new with given uuid
+                        tree = createTree(parent, nodeInfo, id);
+                    }
                 }
             }
-        }
 
-        // process properties
-        importProperties(tree, propInfos, false);
+            // process properties
+            importProperties(tree, propInfos, false);
 
-        if (tree.exists()) {
-            parents.push(tree);
+            if (tree.exists()) {
+                EffectiveNodeType parentEnt = effectiveNodeTypeProvider.getEffectiveNodeType(tree.getParent());
+                EffectiveNodeType ent = effectiveNodeTypeProvider.getEffectiveNodeType(tree);
+                // this throws ConstraintViolationException
+                parentEnt.getNodeDefinition(tree.getName(), ent);
+                parents.push(tree);
+            }
+        } catch (ConstraintViolationException e) {
```

Comment:
only this part is actually new

Reply:
would it be possible in this case to change to PR such that the changes are just limited to what really needs to change?

Reply:
Wrapping with a try...catch requires changing indentation. Only the diff view from Github is not able to correctly visualize the changes in a clever way... As I said above, I have not changed anything apart from that try...catch.